### DISCUSSION TITLE
Cell segmentation parallelization

### DISCUSSION THREAD

**Discussion by quentinblampey** [gustaveroussy/sopa#36] [General]
For efficiency, cell segmentation must be parallelized. This is done automatically using the [Snakemake pipeline](https://gustaveroussy.github.io/sopa/tutorials/snakemake/), but it's possible to use your own parallelization mechanism if you prefer using the [CLI](https://gustaveroussy.github.io/sopa/tutorials/cli_usage/) or the [API](https://gustaveroussy.github.io/sopa/tutorials/api_usage/).

Note that Snakemake may require setting up a profile. There is already an existing one for Slurm clusters, but some other profiles can be added in the future (don't hesitate to ask for it!).

You can ask questions here, or show your parallelization technique to help other users.

**Comment by Marius1311** [comment]
An easy, and maybe a bit naive, way of doing this with dask through the python API would be

```Python
from dask.distributed import Client
from dask import delayed
import dask

# Initialize a Dask client to manage distributed workers
client = Client(n_workers=16)  # Adjust based on the number of patches you have. 

# Define a delayed function to process each patch
@delayed
def process_patch_delayed(patch_index):
    segmentation.write_patch_cells(cellpose_temp_dir, patch_index)

# Wrap patch processing in delayed objects
tasks = [process_patch_delayed(patch_index) for patch_index in range(len(sdata["sopa_patches"]))]

# Compute in parallel using Dask
dask.compute(*tasks)
client.close()
```

Works well in one example case, I haven't tested this extensively. Also, I could not get this to work with Baysor yet - it always runs in sequential manner. You can use the `Client` to monitor performance trough the dask dashboard.

**Reply by quentinblampey** [reply]
Thanks @Marius1311, it looks very interesting, I'll try it one day!

In the future, I may try to benchmark different parallelization techniques, and add the best techniques in the API

**Reply by Marius1311** [reply]
Sounds great! Have you tried parallelizing Baysor through the Python API as well? I could not get that to run yet - whatever I tried, it ended up using a single process only.

**Reply by quentinblampey** [reply]
No, I never tried, I always parallelize Baysor using multiple jobs on a Slurm cluster.
But I think this is a known issue of Bayser, and if I remember correctly, its maintainer may be working on this!

**Reply by lguerard** [reply]
I managed to get Baysor running multithreaded (as far as I understand it though...) with these lines.

```python
def run_baysor(x, baysor_exec_path, baysor_temp_dir, thread_number):
    import os
    command = (
        f"{baysor_exec_path} run --save-polygons GeoJSON -c {baysor_temp_dir}/{x}/config.toml {baysor_temp_dir}/{x}/transcripts.csv -o {baysor_temp_dir}/{x} :cell",
    )
    joined_command = " && ".join(command)
    os.system(joined_command)

import sys
from functools import partial 
from p_tqdm import p_umap

if sys.platform == "win32" and os.cpu_count() > 61:
    num_threads = 61
else:
    num_threads = None

result_objects = p_umap(partial(run_baysor, baysor_exec_path = BAYSOR_EXECUTABLE_PATH, baysor_temp_dir = baysor_temp_dir, thread_number = num_threads), valid_indices, num_cpus=num_threads)
```

On the other side, your code works for me but when running it in a notebook cell, it sends an error message that it's OOM on the GPU (single GPU, 11GB VRAM), the cell is considered as finished executing but processes are still happening and new parquet files get created. Any ideas on how to improve this ? Sorry I don't know much about `dask`...

**Comment by quentinblampey** [comment]
I'm working on `sopa==2.0.0`, which will significantly improve and ease the API usage. It will notably include some parallelization backends, such as dask (adapted from @Marius1311 comment), and maybe others in the future.

Example usage:
```python

sopa.setting.parallelization_backend = "dask" # sets globally the backend

sopa.segmentation.cellpose(sdata, "DAPI", diameter=35) # runs cellpose on all patches in parallel
```

What do you think?

**Reply by Marius1311** [reply]
Looks great!

**Reply by stergioc** [reply]
Hi @quentinblampey, just jumping here quickly to mention that such functionality would be also very useful on other api functions like: `sopa.patches.infer.infer_wsi_patches`.

**Reply by quentinblampey** [reply]
Thanks @stergioc, I can have a look! It should be very easy to change, since internally we only need to do `sopa.settings._run_with_backend(functions)`, where functions is a list of callable.

**Comment by KunHHE** [comment]
Thanks all! If in the case of running parallelization if we prefer using the [CLI], will the sopa==2.00 add parallelization using CLI?

**Reply by quentinblampey** [reply]
Yes @KunHHE the parallelization will also be available for the CLI

**Comment by kristiajazi** [comment]
Hi, 
I am using the CLI for segmenting xenium images , but I feel that the pipeline is running slow even with the parallelization. 
Here are my commands , it would be great to get some input. Thank you!

pip install sopa [cellpose]
sopa convert sopa_directory --technology xenium
set SOPA_PARALLELIZATION_BACKEND=dask
sopa patchify image sopa_directory.zarr 
sopa segmentation cellpose sopa_directory.zarr --diameter 30 --channels DAPI --gpu

even if I make smaller patches , the segmentation takes more than 24 h.

**Reply by quentinblampey** [reply]
Hi @kristiajazi,

I have a few questions for you:
- Do you use cellpose v4? It's a much bigger model than the models from cellpose v3, so it may be slow even on GPUs. Therefore, you could consider trying cellpose v3. Usually, v3 is fast even on CPUs.
- Is there a reason for not using the Snakemake or the (experimental) Nextflow pipeline? It will distribute one patch per job, and it's usually a better solution for parallelization. Then `dask` backend is usually preferred for the API usage, although it's also available via the CLI.
- I see you have Xenium data, so I just wanted to double-check with you if running cellpose was intended? Usually, on Xenium data, the prior 10X multimodal segmentation is a good prior for transcript-based segmentation (such as Proseg). But I know that sometimes, depending on your data/setup, the prior segmentation may not be good enough and you may want to combine Cellpose + Proseg within Sopa.

**Reply by kristiajazi** [reply]
Hi!

1)Yes, I can downgrade to v3, it is a good point.
2) I found the CLI more straightforward to run and when running it , a warning message suggested the dask backend. When I use snakemake I get this error "write_patch_cells-714f3aa5-4611-4942-8585-e61d1947e64c on 4 different workers, but all those workers died while running it", which is something I am trying to solve in parallel. 
3) Yes, it is intended to run Cellpose. My samples did not have the staining for the 10x multimodal segmentation. They were processed before 10x introduced that method.

Thank you for your answer!

**Reply by quentinblampey** [reply]
Alright, sounds good, thanks for your answer too!

**Reply by kristiajazi** [reply]
You are welcome!

The v3 worked better indeed.

But I realized that the issue was actually connected to the error that I got also using snakemake. The problem was that the sopa config.yaml file for xenium suggests "patch_width_pixel" of 6000, which is too big as patch width, it crashes the run both on supercomputers and regular laptops (32,0 GB RAM,64-bit operating system), I tried both . I had to use 2000. Probably, a note about it is required, so other xenium users know it.

Bests,
Kristi

**Reply by quentinblampey** [reply]
Thanks for the feedback, I'll update the config to use 2000 by default!
